api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/AbstractSeparableConv.kt

```diff
@@ -0,0 +1,207 @@
+                multiply(*stridesInternal).toDouble()).roundToInt()
+
+        val (depthwiseKernelVariableName, pointwiseKernelVariableName, biasVariableName) = defineVariableNames()
+
+        createSeparableConvVariables(
+            tf,
+            depthwiseKernelVariableName,
+            pointwiseKernelVariableName,
+            biasVariableName,
+            kGraph
+        )
+    }
+
+    override fun computeOutputShape(inputShape: Shape): Shape {
+        val shape = defineOutputShape(inputShape)
+        outputShape = TensorShape(shape)
+        return shape
+    }
+
+    override fun forward(
+        tf: Ops,
+        input: Operand<Float>,
+        isTraining: Operand<Boolean>,
+        numberOfLosses: Operand<Float>?
+    ): Operand<Float> {
+        var output = separableConvImplementation(tf, input)
+
+        if (useBiasInternal) {
+            output = tf.nn.biasAdd(output, bias)
+        }
+
+        return Activations.convert(activationInternal).apply(tf, output, name)
+    }
+
+    private fun defineVariableNames(): Triple<String, String, String> {
+        return if (name.isNotEmpty()) {
+            Triple(
+                depthwiseKernalVarName(name),
+                pointwiseKernelVarName(name),
+                biasVarName(name)
+            )
+        } else {
+            Triple(depthwiseKernelVariableName, pointwiseKernelVariableName, biasVariableName)
+        }
+    }
+
+    private fun createSeparableConvVariables(
+        tf: Ops,
+        depthwiseKernelVariableName: String,
+        pointwiseKernelVariableName: String,
+        biasVariableName: String,
+        kGraph: KGraph
+    ) {
+        depthwiseKernel = tf.withName(depthwiseKernelVariableName).variable(depthwiseKernelShape, getDType())
+        pointwiseKernel = tf.withName(pointwiseKernelVariableName).variable(pointwiseKernelShape, getDType())
+        if (useBiasInternal) bias = tf.withName(biasVariableName).variable(biasShape, getDType())
+
+        depthwiseKernel = addWeight(
+            tf,
+            kGraph,
+            depthwiseKernelVariableName,
+            depthwiseKernel,
+            depthwiseInitializerInternal,
+            depthwiseRegularizerInternal
+        )
+        pointwiseKernel = addWeight(
+            tf,
+            kGraph,
+            pointwiseKernelVariableName,
+            pointwiseKernel,
+            pointwiseInitializerInternal,
+            pointwiseRegularizerInternal
+        )
+        if (useBiasInternal)
+            bias = addWeight(tf, kGraph, biasVariableName, bias!!, biasInitializerInternal, biasRegularizerInternal)
+    }
+
+    protected open fun getInputDepth(numberOfChannels: Long): Long = numberOfChannels
+
+    protected open fun getOutputDepth(numberOfChannels: Long): Long = filtersInternal
+
+    private fun computeMatricesShapes(numberOfChannels: Long) {
+        depthwiseKernelShape = shapeFromDims(*kernelSizeInternal, numberOfChannels, depthMulitplierInternal.toLong())
```

Comment:
maybe better to make the function without side-effects, which returns a triple of shapes